pkg/composableschemadsl/compiler/translator.go

```diff
@@ -696,27 +742,167 @@ func addWithCaveats(tctx translationContext, typeRefNode *dslNode, ref *core.All
+	// of imported nodes
+	importedDefinitionNodes := list.New()
+
+	for _, topLevelNode := range root.GetChildren() {
+		// Process import nodes; ignore the others
+		if topLevelNode.GetType() == dslshape.NodeTypeImport {
+			// Do the handling of recursive imports etc here
+			importPath, err := topLevelNode.GetString(dslshape.NodeImportPredicatePath)
+			if err != nil {
+				return err
+			}
+
+			if err := validateFilepath(importPath); err != nil {
+				return err
+			}
+			filePath := filepath.Join(itctx.sourceFolder, importPath)
+
+			newSourceFolder := filepath.Dir(filePath)
+
+			currentLocallyVisitedFiles := itctx.locallyVisitedFiles.Copy()
+
+			if ok := currentLocallyVisitedFiles.Add(filePath); !ok {
+				// If we've already visited the file on this particular branch walk, it's
+				// a circular import issue.
+				return &CircularImportError{
+					error:    fmt.Errorf("circular import detected: %s has been visited on this branch", filePath),
+					filePath: filePath,
+				}
+			}
+
+			if ok := itctx.globallyVisitedFiles.Add(filePath); !ok {
+				// If the file has already been visited, we short-circuit the import process
+				// by not reading the schema file in and compiling a schema with an empty string.
+				// This prevents duplicate definitions from ending up in the output, as well
+				// as preventing circular imports.
+				log.Debug().Str("filepath", filePath).Msg("file %s has already been visited in another part of the walk")
+				return nil
+			}
+
+			// Do the actual import here
+			// This is a new node provided by the translateImport
+			parsedImportRoot, err := importFile(filePath)
+			if err != nil {
+				return err
+			}
+
+			// We recurse on that node to resolve any further imports
+			err = translateImports(importTranslationContext{
+				sourceFolder:         newSourceFolder,
+				locallyVisitedFiles:  currentLocallyVisitedFiles,
+				globallyVisitedFiles: itctx.globallyVisitedFiles,
+			}, parsedImportRoot)
+			if err != nil {
+				return err
+			}
+
+			// And then append the definition to the list of definitions to be added
+			for _, importedNode := range parsedImportRoot.GetChildren() {
+				if importedNode.GetType() != dslshape.NodeTypeImport {
+					importedDefinitionNodes.PushBack(importedNode)
+				}
+			}
+		}
+	}
+
+	// finally, take the list of definitions to add and prepend them
+	// (effectively hoists definitions)
+	root.ConnectAndHoistMany(dslshape.NodePredicateChild, importedDefinitionNodes)
+
+	return nil
+}
+
+func translatePartials(tctx translationContext, rootNode *dslNode) error {
+	for _, topLevelNode := range rootNode.GetChildren() {
+		if topLevelNode.GetType() == dslshape.NodeTypePartial {
+			err := translatePartial(tctx, topLevelNode)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	if tctx.partialWaitingRooms.Len() != 0 {
+		return fmt.Errorf("could not resolve partials: [%s]. this may indicate a circular reference", strings.Join(tctx.partialWaitingRooms.Keys(), ", "))
 	}
```

Comment:
This engenders an assumption that the algorithm using this approach is correct. So far it looks correct, but I don't know that I could prove it right now.